### ISSUE TITLE
Transmutes from non repr(C) to repr(C) structs

### ISSUE THREAD

**Issue by nagisa** [servo/rust-azure#250]
For example this:

https://github.com/servo/rust-azure/blob/master/src/azure_hl.rs#L708

```
    pub fn draw_filter(&self,
                       filter: &FilterNode,
                       source_rect: &Rect<AzFloat>,
                       dest_point: &Point2D<AzFloat>,
                       options: DrawOptions) {
        unsafe {
            AzDrawTargetDrawFilter(self.azure_draw_target,
                                   filter.azure_filter_node,
                                   mem::transmute::<_,*const AzRect>(source_rect),
                                   mem::transmute::<_,*const AzPoint>(dest_point),
                                   &options.as_azure_draw_options())
        }
}
```

transmutes from `&Rect<AzFloat>` to `*const AzRect`.

`AzRect` is an alias to `repr(C)` type, but `Rect<AzFloat>` is not:

```rust
#[derive(RustcDecodable, RustcEncodable)]
pub struct TypedRect<T, U = UnknownUnit> {
    pub origin: TypedPoint2D<T, U>,
    pub size: TypedSize2D<T, U>,
}

/// The default rectangle type with no unit.
pub type Rect<T> = TypedRect<T, UnknownUnit>;
```

Therefore this whole crate is invoking a big undefined behaviour and should probably be fixed, preferably by not using transmute.

**Comment by Ms2ger** [comment]
This was mostly fixed by #251, but not entirely. The following are left:

```rust
    pub fn create_gradient_stops(&self,
                                 gradient_stops: &[GradientStop],
                                 extend_mode: ExtendMode)
                                 -> GradientStops {
        unsafe {
            GradientStops::new(AzDrawTargetCreateGradientStops(
                    self.azure_draw_target,
                    mem::transmute::<_,*const AzGradientStop>(gradient_stops.as_ptr()),
                    gradient_stops.len() as u32,
                    extend_mode.as_azure_extend_mode()))
        }
    }

    pub fn set_transform(&self, matrix: &Matrix2D<AzFloat>) {
        unsafe {
            AzDrawTargetSetTransform(self.azure_draw_target, mem::transmute(matrix));
        }
    }
```

CC @nox

**Comment by nox** [comment]
@Ms2ger The first one is ok because `GradientStop` is actually `repr(C)`. The second one though, clearly an oversight on my part.
